console-subscriber/src/aggregator.rs

```diff
@@ -615,9 +906,287 @@ impl Task {
+                    let new_val = *val + upd;
+                    *val = new_val;
+                }
+                AttributeUpdateOp::Ovr => {
+                    *val = *upd;
+                }
+                AttributeUpdateOp::Sub => {
+                    let new_val = *val - upd;
+                    *val = new_val;
+                }
+            },
+            _ => panic!("trying to update attributes of different type"),
+        }
+    }
+}
+
+impl From<AttributeUpdateValue> for AttrValue {
+    fn from(upd: AttributeUpdateValue) -> Self {
+        match upd {
+            AttributeUpdateValue::Text(t) => AttrValue::Text(t),
+            AttributeUpdateValue::Numeric { val, unit, .. } => AttrValue::Numeric { val, unit },
+        }
+    }
+}
+
 fn serialize_histogram(histogram: &Histogram<u64>) -> Result<Vec<u8>, V2SerializeError> {
     let mut serializer = V2Serializer::new();
     let mut buf = Vec::new();
     serializer.serialize(histogram, &mut buf)?;
     Ok(buf)
 }
+
+fn total_time(created_at: Option<&SystemTime>, closed_at: Option<&SystemTime>) -> Option<Duration> {
+    closed_at.and_then(|end| created_at.and_then(|start| end.duration_since(*start).ok()))
+}
+
+/// Drops all tasks, resources and ops that are not alive anymore
+fn drop_closed<T, R: Closable>(
+    now: SystemTime,
+    entities: &mut IdData<T>,
+    stats: &mut IdData<R>,
+    retention: Duration,
+    has_watchers: bool,
+) {
+    // drop stats for closed tasks if they have been updated
+    tracing::trace!(?retention, has_watchers, "dropping closed entities...");
+
+    let stats_len_0 = stats.data.len();
+    stats.data.retain(|id, (stats, dirty)| {
+        if let Some(closed) = stats.closed_at() {
+            let closed_for = now.duration_since(*closed).unwrap_or_default();
+            let should_drop =
+                    // if there are any clients watching, retain all dirty tasks regardless of age
+                    (*dirty && has_watchers)
+                    || closed_for > retention;
+            tracing::trace!(
+                stats.id = ?id,
+                stats.closed_at = ?closed,
+                stats.closed_for = ?closed_for,
+                stats.dirty = *dirty,
+                should_drop,
+            );
+            return !should_drop;
+        }
+
+        true
+    });
+
+    let stats_len_1 = stats.data.len();
+
+    // drop closed entities which no longer have stats.
+    let entities_len_0 = entities.data.len();
+    entities
+        .data
+        .retain(|id, (_, _)| stats.data.contains_key(id));
+    let entities_len_1 = entities.data.len();
+    let dropped_stats = stats_len_0 - stats_len_1;
+
+    let stats_len_1 = stats.data.len();
+    if dropped_stats > 0 {
+        tracing::debug!(
+            tasks.dropped = entities_len_0 - entities_len_1,
+            tasks.len = entities_len_1,
+            stats.dropped = dropped_stats,
+            stats.tasks = stats_len_1,
+            "dropped closed entities"
```

Comment:
nit/tioli: should maybe also be 
```suggestion
            "dropped closed {}",
            std::any::type_name::<R>(),
```